// shieldbow/dist/structures/api/ChallengeRank.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.ChallengeRank = void 0;
/**
 * A representation of the ranked challenges stats of a summoner.
 */
class ChallengeRank {
    constructor(client, data, tier) {
        /**
         * The unique player ID of the summoner.
         */
        Object.defineProperty(this, "playerId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The tier this summoner belongs to regarding this challenge.
         */
        Object.defineProperty(this, "tier", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The progression of the summoner in the challenge.
         */
        Object.defineProperty(this, "progression", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The rank of the summoner in the challenge.
         */
        Object.defineProperty(this, "rank", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The client this manager belongs to.
         */
        Object.defineProperty(this, "client", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.client = client;
        this.playerId = data.puuid;
        this.tier = tier;
        this.progression = data.value;
        this.rank = data.position;
    }
    /**
     * Fetch the details of the summoner.
     */
    fetchSummoner() {
        return this.client.summoners.fetchByPlayerId(this.playerId);
    }
}
exports.ChallengeRank = ChallengeRank;
