// shieldbow/dist/structures/api/CurrentGameTeam.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CurrentGameTeam = void 0;
const CurrentGameParticipant_1 = require("./CurrentGameParticipant");
/**
 * A representation of a team in a live game.
 */
class CurrentGameTeam {
    /**
     * Creates a new Current Game Team instance.
     * @param client - The client requesting the data.
     * @param bans - The raw bans data for this team from the API.
     * @param participants - The raw participants data for this team from the API.
     * @param champions - The champions involved in the game.
     * @param runeTrees - The collection of the runes in the game.
     * @param summonerSpells - The collection of the summoner spells in the game.
     */
    constructor(client, bans, participants, champions, runeTrees, summonerSpells) {
        /**
         * The team's ID.
         */
        Object.defineProperty(this, "id", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The team's side.
         */
        Object.defineProperty(this, "side", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The champions banned by this team.
         */
        Object.defineProperty(this, "bans", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The team's participants.
         */
        Object.defineProperty(this, "participants", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.id = participants[0].teamId;
        this.side = this.id === 100 ? 'blue' : 'red';
        this.bans = bans.map((b) => ({
            champion: champions.find((c) => c.key === b.championId),
            turn: b.pickTurn
        }));
        this.participants = participants.map((p) => new CurrentGameParticipant_1.CurrentGameParticipant(client, p, champions.find((c) => c.key === p.championId), runeTrees, summonerSpells));
    }
}
exports.CurrentGameTeam = CurrentGameTeam;
