// shieldbow/dist/structures/api/MatchTimeline.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.MatchTimeline = void 0;
const TimelineFrame_1 = require("./TimelineFrame");
/**
 * A representation of the timeline data for a match.
 */
class MatchTimeline {
    /**
     * Creates a new match timeline instance.
     *
     * @param data - The raw timeline data from the API.
     * @param items - A collection of all items in the game.
     */
    constructor(data, items) {
        /**
         * The version of the timeline data.
         */
        Object.defineProperty(this, "dataVersion", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The ID of the match.
         */
        Object.defineProperty(this, "matchId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The interval between frames.
         */
        Object.defineProperty(this, "frameInterval", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The IDs of the participants in the match.
         */
        Object.defineProperty(this, "participantIds", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The returned frames in the match timeline.
         */
        Object.defineProperty(this, "frames", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.dataVersion = data.metadata.dataVersion;
        this.matchId = data.metadata.matchId;
        this.participantIds = data.metadata.participants;
        this.frameInterval = data.info.frameInterval;
        this.frames = data.info.frames.map((frame) => new TimelineFrame_1.TimelineFrame(frame, items));
    }
}
exports.MatchTimeline = MatchTimeline;
