// shieldbow/dist/structures/api/Team.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Team = void 0;
const Participant_1 = require("./Participant");
/**
 * A representation of a team in a match.
 */
class Team {
    constructor(data, participants, champions, items, runeTrees, summonerSpells) {
        /**
         * The ID of the team.
         */
        Object.defineProperty(this, "id", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The champions banned by the team.
         */
        Object.defineProperty(this, "bans", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The objectives secured by the team.
         */
        Object.defineProperty(this, "objectives", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * Whether the team won the match.
         */
        Object.defineProperty(this, "win", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The participants in the team.
         */
        Object.defineProperty(this, "participants", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.id = data.teamId;
        this.bans = data.bans.map((b) => ({
            order: b.pickTurn,
            champion: champions.find((c) => c.key === b.championId)
        }));
        this.objectives = data.objectives;
        this.participants = participants.map((p) => new Participant_1.Participant(p, champions.find((c) => c.key === p.championId), items, runeTrees, summonerSpells));
        this.win = data.win;
    }
}
exports.Team = Team;
