// shieldbow/dist/structures/api/TournamentTeam.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TournamentTeam = void 0;
const TournamentPlayer_1 = require("./TournamentPlayer");
/**
 * A representation of a team in a clash tournament.
 */
class TournamentTeam {
    /**
     * Creates a new tournament team instance.
     * @param client - The client that requested this data.
     * @param data - The raw tournament team data from the API.
     */
    constructor(client, data) {
        /**
         * The ID of the team.
         */
        Object.defineProperty(this, "id", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The ID of the tournament this team is participating in.
         */
        Object.defineProperty(this, "tournamentId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The name of the team.
         */
        Object.defineProperty(this, "name", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The icon being used to represent the team.
         */
        Object.defineProperty(this, "icon", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The tier of the team.
         */
        Object.defineProperty(this, "tier", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The summoner ID of the team's captain.
         */
        Object.defineProperty(this, "captainId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The abbreviation for the team's name.
         */
        Object.defineProperty(this, "abbreviation", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The list of players in the team.
         */
        Object.defineProperty(this, "players", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.id = data.id;
        this.tournamentId = data.tournamentId;
        this.name = data.name;
        this.icon = `${client.cdnBase}${client.version}/img/profileicon/${data.iconId}.png`;
        this.tier = data.tier;
        this.captainId = data.captain;
        this.abbreviation = data.abbreviation;
        this.players = data.players.map((player) => new TournamentPlayer_1.TournamentPlayer(client, this.id, player));
    }
}
exports.TournamentTeam = TournamentTeam;
