// shieldbow/dist/structures/api/TournamentSchedule.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TournamentSchedule = void 0;
/**
 * A representation of a clash tournament schedule.
 */
class TournamentSchedule {
    /**
     * Creates a new tournament schedule instance.
     * @param data - The raw tournament schedule data from the API.
     */
    constructor(data) {
        /**
         * The ID of the tournament schedule.
         */
        Object.defineProperty(this, "id", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The timestamp at which the registration begins.
         */
        Object.defineProperty(this, "registrationTimestamp", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The timestamp at which the tournament begins.
         */
        Object.defineProperty(this, "startTimestamp", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * Whether the events as per this schedule have been cancelled.
         */
        Object.defineProperty(this, "cancelled", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.id = data.id;
        this.registrationTimestamp = data.registrationTime;
        this.startTimestamp = data.startTime;
        this.cancelled = data.cancelled;
    }
    /**
     * The time at which the registration begins.
     */
    get registrationTime() {
        return new Date(this.registrationTimestamp);
    }
    /**
     * The time at which the tournament begins.
     */
    get startTime() {
        return new Date(this.startTimestamp);
    }
}
exports.TournamentSchedule = TournamentSchedule;
