// shieldbow/dist/structures/api/Tournament.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.Tournament = void 0;
const TournamentSchedule_1 = require("./TournamentSchedule");
/**
 * A representation of a clash tournament.
 */
class Tournament {
    /**
     * Creates a new tournament instance.
     * @param data - The raw tournament data from the API.
     */
    constructor(data) {
        /**
         * The ID of the tournament.
         */
        Object.defineProperty(this, "id", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The ID of the theme of the tournament.
         */
        Object.defineProperty(this, "themeId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The name (title) of the tournament.
         */
        Object.defineProperty(this, "title", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The description (day) of the tournament.
         */
        Object.defineProperty(this, "subtitle", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The tournament's schedule.
         */
        Object.defineProperty(this, "schedule", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.id = data.id;
        this.themeId = data.themeId;
        this.title = data.nameKey;
        this.subtitle = data.nameKeySecondary;
        this.schedule = data.schedule.map((s) => new TournamentSchedule_1.TournamentSchedule(s));
    }
}
exports.Tournament = Tournament;
