// shieldbow/dist/structures/api/TournamentPlayer.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TournamentPlayer = void 0;
/**
 * A representation of a player in a clash tournament.
 */
class TournamentPlayer {
    /**
     * Creates a new tournament player instance.
     * @param client - The client that requested this data.
     * @param teamId - The ID of the team this player is on.
     * @param data - The raw player data from the API.
     */
    constructor(client, teamId, data) {
        /**
         * The ID of the team this player is on.
         */
        Object.defineProperty(this, "teamId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The summoner ID of this player.
         */
        Object.defineProperty(this, "summonerId", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The role this player will be performing in the team - CAPTAIN or MEMBER.
         */
        Object.defineProperty(this, "role", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        /**
         * The position this player will be playing in the team.
         */
        Object.defineProperty(this, "position", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        Object.defineProperty(this, "client", {
            enumerable: true,
            configurable: true,
            writable: true,
            value: void 0
        });
        this.client = client;
        this.teamId = teamId;
        this.summonerId = data.summonerId;
        this.role = data.role;
        this.position = data.position;
    }
    /**
     * Use {@link SummonerManager.fetch} to fetch the summoner for this player.
     */
    get summoner() {
        return this.client.summoners.fetch(this.summonerId);
    }
}
exports.TournamentPlayer = TournamentPlayer;
