// shieldbow/dist/ratelimiter/constants.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.methods = void 0;
/**
 * The API method names.
 */
exports.methods = {
    ACCOUNT: ['getByPuuid', 'getByRiotId'],
    CHAMPION_MASTERY: [
        'getAllChampionMasteries',
        'getChampionMastery',
        'getTopChampionMasteries',
        'getChampionMasteryScore'
    ],
    CHAMPION: ['getChampionInfo'],
    CLASH: ['getTeamById', 'getPlayersBySummoner', 'getTournaments', 'getTournamentById', 'getTournamentByTeam'],
    LEAGUE_EXP: ['getLeagueEntries'],
    LEAGUE: ['getLeagueEntriesForSummoner', 'getLeagueById'],
    LOL_CHALLENGES: [
        'getAllChallengeConfigs',
        'getAllChallengePercentiles',
        'getChallengeConfigs',
        'getChallengeLeaderboards',
        'getChallengePercentiles',
        'getPlayerData'
    ],
    LOL_STATUS: ['getPlatformData'],
    MATCH: ['getMatchIdsByPUUID', 'getMatch', 'getTimeline'],
    SPECTATOR: ['getCurrentGameInfoBySummoner', 'getFeaturedGames'],
    SUMMONER: ['getByAccountId', 'getBySummonerName', 'getByPUUID', 'getBySummonerId']
};
