// shieldbow/dist/structures/dragon/index.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SummonerSpell = exports.RuneTree = exports.Rune = exports.Item = exports.ChampionSpell = exports.ChampionSkin = exports.ChampionStat = exports.Champion = void 0;
var Champion_1 = require("./Champion");
Object.defineProperty(exports, "Champion", { enumerable: true, get: function () { return Champion_1.Champion; } });
var ChampionStat_1 = require("./ChampionStat");
Object.defineProperty(exports, "ChampionStat", { enumerable: true, get: function () { return ChampionStat_1.ChampionStat; } });
var ChampionSkin_1 = require("./ChampionSkin");
Object.defineProperty(exports, "ChampionSkin", { enumerable: true, get: function () { return ChampionSkin_1.ChampionSkin; } });
var ChampionSpell_1 = require("./ChampionSpell");
Object.defineProperty(exports, "ChampionSpell", { enumerable: true, get: function () { return ChampionSpell_1.ChampionSpell; } });
var Item_1 = require("./Item");
Object.defineProperty(exports, "Item", { enumerable: true, get: function () { return Item_1.Item; } });
var Rune_1 = require("./Rune");
Object.defineProperty(exports, "Rune", { enumerable: true, get: function () { return Rune_1.Rune; } });
var RuneTree_1 = require("./RuneTree");
Object.defineProperty(exports, "RuneTree", { enumerable: true, get: function () { return RuneTree_1.RuneTree; } });
var SummonerSpell_1 = require("./SummonerSpell");
Object.defineProperty(exports, "SummonerSpell", { enumerable: true, get: function () { return SummonerSpell_1.SummonerSpell; } });
